[Major Changes][Sprockets 3] Supporting multiple sprockets versions and using versionized classes.

bogdanRada opened the pull request:
Hello , as discussed here: https://github.com/petebrowne/sprockets-sass/pull/38 i made the changes for using different classes for different versions. 

**Tilt dependency is still removed**, but if you want to keep it for the legacy versions , let me know and i can rewrite the V2 class for SASS template. 

All tests are passing. for botth sprockets 2.x and 3.x, however sprockets 4 is not support yet, becasue i need to fix the importer for that version

Also fixed all issues i could find with Rubocop. 

**THIS WILL ALSO DROP Ruby 1.9** 

Let me know if this is better now :)

bogdanRada commented:
Of course, some more cleaning can be done,if needed. Let me know if you want to refactor something or change anything. Thanks

bogdanRada commented:
I think we should also update the Sprockets::Sass::V3::Functions to be the same as the ones from Sprockets-herlpers, some of them were changed to require two parameters, But i am not sure if we need to keep compatibility with that gem too.

This functions will need to get refactored though for sprockets 4, becauase context.logical_path is now throwing NoMethodError. 

Just wanted to know if we want to keep that compatibility or not.

petebrowne commented on lib/sprockets/sass/registration.rb:
```diff
@@ -0,0 +1,141 @@
+      end
+
+      def require_libraries
+        AVAILABLE_VERSIONS.each do |version|
```
Shouldn't this only require the the libraries for the current version of sprockets? If we're requiring all of them, we could just explicitly list the require statements

bogdanRada replied:
Usually yes, but considering for example for sprockets 4....from the list of files only three of them are currently defined for that version....the other ones will be used from v3...since no changes are there.

petebrowne commented on lib/sprockets/sass/registration.rb:
```diff
@@ -0,0 +1,141 @@
+        end
+      end
+
+      def action_to_be_executed
```
I think it'd be easier to understand what's going on with a simple case_statement. something like:

``` ruby
case @sprockets_version
when 2..3
  register_sprockets_legacy
when 3..4
  register_sprockets_v3
when 4..5
  register_sprockets_v4
else
  raise "Version #{@sprockets_version} is not supported"
end
```

bogdanRada replied:
ok, can do that too...jus that didn't wanted to become very long over time.

petebrowne commented on lib/sprockets/sass/registration.rb:
```diff
@@ -0,0 +1,141 @@
+          _register_transformers(from: "application/#{mime}+ruby", to: "text/#{mime}", klass: Sprockets::ERBProcessor)
+        end
+        _register_v4_preprocessors(
+          Sprockets::Sass::Utils.get_class_by_version('SassTemplate') => ['text/sass', 'text/css'],
```
Could you just use `Sprockets::Sass::V4::SassTemplate`?

bogdanRada replied:
I could , but the reason why it is like that, is that if a greater version will be released, if nothing changes for that version, this will use the last version that is higher.   I can do that though if you want . Just that you'll have to change that again if you will need to use another version. 

So this will automatically pick the higher version available. But if you think hardcoding it is better, i am fine with that.

petebrowne commented on lib/sprockets/sass/v3/functions.rb:
```diff
@@ -0,0 +1,142 @@
+  module Sass
+    module V3
+      # Module used to inject helpers into SASS engine
+      module Functions
```
Does this change at all from V2? Should this just `include V2::Functions`?

bogdanRada replied:
Currently there are no changes, but i copied it because i wanted to ask about compatibility with sprockets-helpers.  If we need to maintain that compatibility, i have to refactor this functions. 

If not, i can just delete it because the method ```get_class_by_version``` will automatically pick the highest version available.

petebrowne replied:
It'd be nice to maintain compatibility with sprockets-helpers - but that could be done in a separate PR if it's a challenge

petebrowne commented on lib/sprockets/sass/v2/importer.rb:
```diff
@@ -0,0 +1,205 @@
+          Pathname.new path.to_s.sub(/([^\/]+)\Z/, '_\1')
+        end
+
+        def opposite_syntax(path)
```
Where is this used?

bogdanRada replied:
The partialize_path is used in method ```possible_files```

petebrowne replied:
I was referring to the `opposite_syntax` method - I don't see where that's being used

bogdanRada replied:
Ahhh, i forgot about it. I removed it just now :) Thanks

bogdanRada commented:
I updated the code as you suggested, all tests are passsing, I am still working on V4 Importer.

bogdanRada commented:
I improved a bit the v4 importer....now only 9 failures are there...but i am still working on the other failures.

bogdanRada commented:
I hope now is better, I can rebase and squash the commits if you approve the changes. Let me know if i need to change something else . THanks

petebrowne commented:
Thanks, this is really great. What's left todo?

[] sprockets-helpers compatibility
[] sprockets 4 compatibility

^ Both of these could be separate PRs

bogdanRada commented:
Yes, i agree, and some real life testing on applications, to see if this is working as expected. 

I rebased the commits and squashed them into a single commit. 


Also i have a question, seems that in newer versions of sass , is possible to use ```render_with_sourcemap``` . Do you want to add some support for rendering with Sourcemap? 
That can be done in a separate PR also if needed. 

Let me know if something else is needed from this PR, refactoring something or changing anything. Thank you very much for your support.

bogdanRada commented:
I tested with older version of sprockets 2.12 and some other versions (2.x) , they seem to work fine. However i don't have an application with Sprockets 3.7 ready . I can try and test with latest version also but will need some time to prepare an application for that if is necessary.

petebrowne commented:
👍 Merging this in - I'm planning to release this as 2.0.0.beta1 so we can have people test it out in real world applications while we iron out some of the other details.

bogdanRada commented:
Wow....thanks very much :) This is awesome :D
